internal/container/layer.go

```diff
@@ -0,0 +1,254 @@
+	"oras.land/oras-go/v2/errdef"
+)
+
+// RootFS returns a slice of files opened on the tar files of layers making the
+// root file system of a container image.
+//
+// The function automatically fetches and extracts missing layers using the src
+// store and the list of descriptors representing the compressed image layers.
+//
+// The dst parameter is the directory where the layers are extracted, it must
+// exist or the function will error.
+func RootFS(ctx context.Context, dst string, src oras.ReadOnlyTarget, image *ocispec.Image, layers []ocispec.Descriptor) ([]*os.File, error) {
+	files := make([]*os.File, len(image.RootFS.DiffIDs))
+	defer func() {
+		for _, f := range files {
+			f.Close()
+		}
+	}()
+
+	for {
+		var missingLayers []ocispec.Descriptor
+
+		for i, diffID := range image.RootFS.DiffIDs {
+			if files[i] != nil {
+				continue
+			}
+			// Attempt to open the layer file, if we succeed the layer already
+			// existed in the cache directory and there is no need to fetch it
+			// from the object store.
+			f, err := os.Open(layerPath(dst, diffID))
+			if err != nil {
+				if errors.Is(err, fs.ErrNotExist) {
+					missingLayers = append(missingLayers, layers[i])
+					continue
+				}
+				return nil, err
+			}
+			files[i] = f
+		}
+
+		if len(missingLayers) == 0 {
+			ret := files
+			files = nil // prevents the defer from closing the files
+			return ret, nil
+		}
+
+		// Some of the layer files were missing, we need to fetch and extract
+		// them from the object store. On success, the returned list of open
+		// files is guaranteed to have as many files as there were missing
+		// layers, with a 1:1 correspondence on the indexes.
+		missingFiles, err := ExtractLayers(ctx, dst, src, missingLayers)
+		if err != nil {
+			return nil, err
+		}
+
+		j := 0
+		for i, f := range files {
+			if f == nil {
+				files[i] = missingFiles[j]
+				j++
+			}
+		}
+	}
+}
+
+// ExtractLayers fetches the image layers represented by the list of descriptors
+// and extracts them at the dst file path.
+//
+// The content of the layers is written in a directory named after their digest
+// algorithm, and uncompressed in a tar archive named with their digest hash.
+//
+//	(dst)
+//	 └── sha256
+//	    └── 8c6d1654570f041603f4cef49c320c8f6f3e401324913009d92a19132cbf1ac0
+//	        ...
+//
+// The function performs the extraction of each layer concurrently, each layer
+// is processed in a dedicated goroutine.
+//
+// The operation is atomic, either the creation of the layer fully succeeds or
+// it fails and the layer is not present in the directory.
+func ExtractLayers(ctx context.Context, dst string, src oras.ReadOnlyTarget, layers []ocispec.Descriptor) ([]*os.File, error) {
+	files := make([]*os.File, len(layers))
+	errch := make(chan error)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for i, desc := range layers {
+		go func(i int, desc ocispec.Descriptor) {
+			f, err := ExtractLayer(ctx, dst, src, desc)
```

Comment:
Out of curiosity, how many image layers are typical and what's a ballpark estimate for the most layers we might see here? Is it worth limiting concurrency?

Reply:
Usually I'd say less than 5 layers, but it could be more. 

I'll add a concurrency limit to be safe.